django/core/handlers/asgi.py

```diff
@@ -19,6 +20,34 @@
 logger = logging.getLogger('django.request')
 
 
+async def _sync_to_async_iterator(iterator):
+    """
+    Iterate over a sync generator in an async context.
+    """
+    q = asyncio.Queue()
+
+    def sync_iterator():
+        for item in iterator:
+            q.put_nowait(item)
```

Comment:
Am I correct that this approach can result in loading the entire response into memory before starting to send the chunks? If so, it seems like that's a possible downside because it appears to go against [`StreamingHttpResponse`'s purpose](https://docs.djangoproject.com/en/3.2/ref/request-response/#streaminghttpresponse-objects). Without this change, it looks like `send_response()` only has one bytestring of the streaming response's iterator in memory at a time.

Reply:
It won't necessarily - the queue means it'll be unloaded as it's loaded - but I do wonder if giving the queue an explicit length and using normal `put()`, or some similar way of _limiting_ the amount of memory used, would be sensible.

Reply:
It wouldn't necessarily but it _could_ in theory. In practice, I'm not sure.

It's definitely a tradeoff, since calling `put()` would have to use `async_to_sync(put)()` so it would be slower. Although it occurs to me this could still use `put_nowait()` even if there is a `maxsize`, but I'm not sure what to do if `QueueFull` is raised. Sleep for a short amount of time and try again?

Reply:
Hm, yes, I forgot this is an async queue so `put()` won't just sleep synchronously. I think catching QueueFull and sleeping would accomplish the same thing, though, and give you a queue with one sync "end" and one async "end",

Reply:
Thoughts on the amount of time to sleep? Magic numbers are weird. I'll make it 0.1s when I update the PR to use a sentinel value but we can change it later.

Reply:
Also... thoughts on the size of the queue? I feel like setting it to 1 could raise a lot of `QueueFull` errors depending on how the view is written. Would a small queue size like 5 be against the spirit of a streaming response?

Reply:
Well the thing is that what StreamingResponse returns isn't guaranteed to be single bytes - if I remember right, it can be chunks of any size - so the queue size isn't going to directly dictate the number of bytes stored in memory. 5 would be a good first start, but I wonder if there's a way to cap the number of _bytes_ too here, along with the sentinel value idea.